### DISCUSSION TITLE
The refreshable objects are the same under different instances

### DISCUSSION THREAD

**Discussion by Aues6uen11Z** [zauberzeug/nicegui#2965] [Q&A] [answered]
### Question

Hello, I’m encountering a problem while using `@ui.refreshable`. My situation is as follows

I have multiple pages, each displaying content in a different language, and I want to refresh the page immediately when the language setting is changed to switch the displayed language. Here is a simple demo:

```python
from nicegui import app, ui


class SubContent:
    def __init__(self, name, refreshables):
        self.name = name
        self.refreshables = refreshables
    
    def change_lang(self):
        for r in self.refreshables:
            r.refresh()
            print(r)    # The refreshable objects of 'page1' and 'page2' are the same

    def show(self):
        lang = ui.select(['zh-CN', 'en-US'], value='zh-CN').bind_value(app.storage.general, f'{self.name}')
        lang.on_value_change(self.change_lang)

class MainContent:
    def __init__(self, name):
        self.name = name

    @ui.refreshable
    def content1(self):
        lang = app.storage.general.get(self.name, 'zh-CN')
        if lang == 'zh-CN':
            ui.label('你好')
        elif lang == 'en-US':
            ui.label('Hello')

    @ui.refreshable
    def content2(self):
        lang = app.storage.general.get(self.name, 'zh-CN')
        if lang == 'zh-CN':
            ui.label('世界')
        elif lang == 'en-US':
            ui.label('World')
        
        SubContent(self.name, [self.content1, self.content2]).show()
    
    def show(self):
        with ui.tab_panel(self.name):
            with ui.row():
                self.content1()
                self.content2()


with ui.tabs() as tabs:
    ui.tab('page1')
    ui.tab('page2')
with ui.tab_panels(tabs, value='page1') as panels:
    test1 = MainContent('page1').show()
    test2 = MainContent('page2').show()

ui.run()
```

If I have only one page, everything works fine. However, if I create multiple pages, only the last page can switch languages normally. When debugging, I found that the refreshables object passed to the `SubContent` class are the same for different pages. 
![image](https://github.com/zauberzeug/nicegui/assets/79563000/7aa34d79-ce78-46ff-a057-2a65d552e248)
![image](https://github.com/zauberzeug/nicegui/assets/79563000/7c25f652-4258-40d8-bc5a-28d821f15d3e)
I’m not very clear on the underlying principles here. How can I properly set the language for each page individually? Thank you for any possible help.

**Comment by Aues6uen11Z** [comment]
I found the solution. Strangely enough, it seems that “the refreshable object is the same” is a normal phenomenon. I changed the way the `refresh()` was executed and succeeded, but I really don’t understand what principle this works on.

```python
class SubContent:
    ...
    def show(self):
        lang = ui.select(['zh-CN', 'en-US'], value='zh-CN').bind_value(app.storage.general, f'{self.name}')
        for r in self.refreshables:
            lang.on_value_change(r.refresh)
```

**Comment by falkoschindler** [comment, marked as answer]
Hi @Aues6uen11Z,

I think this problem has been discussed in #2535 and boils down to something like this:
```py
class Content:

    def __init__(self):
        with ui.row():
            self.content()
            ui.button('Refresh 1', on_click=lambda refreshable=self.content: refreshable.refresh())
            ui.button('Refresh 2', on_click=lambda: self.content.refresh())
            ui.button('Refresh 3', on_click=self.content.refresh)

    @ui.refreshable
    def content(self):
        ui.label(time.time())

Content()
Content()
```

"Refresh 1" fails to distinguish both content instances, while "Refresh 2" and "Refresh 3" work as expected. This is because when calling `.refresh()` on a class method (which is static), the `refresh` function needs to find out which _instance_ to update. It does this by keeping track of the last instance that has been accessed, i.e. calling `self.content` in this case. So when running the line `self.content.refresh()`, the first part `self.content` takes note of which `self` we're currently talking about, and the last part `.refresh()` uses this information. But if you store multiple references to the result of `self.content` for later, they might be mixed up when actually calling `.refresh()`. This is unfortunate, but we haven't found a better solution yet.

**Reply by Aues6uen11Z** [reply]
I'm sorry for not noticing the previous discussion. Thank you very much for your patient answer！
